src/marshmallow_annotations/builder.py

```diff
@@ -0,0 +1,236 @@
+
+
+class BaseConverter:
+
+    def __init__(self, scheme, *, registry=TypeRegistry):
+        self.registry = TypeRegistry
+        self.scheme = scheme
+
+    def _field_from_typehint(self, typehint, kwargs=None):
+        # need that immutable dict in the stdlib pls
+        kwargs = kwargs if kwargs is not None else {}
+
+        # sane defaults
+        allow_none = False
+        required = True
+
+        if _is_optional(typehint):
+            allow_none = True
+            typehint = typehint.__args__[0]
+
+        # set this after optional check
+        subtypes = getattr(typehint, '__args__', ())
+
+        if subtypes != ():
+            typehint = typehint.__base__
+
+        kwargs.setdefault('allow_none', allow_none)
+        kwargs.setdefault('required', required)
+
+        field_constructor = self.registry.from_type(typehint)
+        return field_constructor(self, subtypes, kwargs)
+
+    def _exists_on_scheme(self, name):
+        if self.scheme is None:
+            return False
+
+    def _get_meta_options(self, name):
+        if self.scheme is None:
+            return {}
+
+        parent_meta_options = self._visit_parent_metas(
+            lambda meta: getattr(meta, name, None)
+        )
+
+        return parent_meta_options or {}
+
+    def _visit_parent_metas(self, f):
+        for parent in self.scheme.__mro__:
+            meta = getattr(parent, 'Meta', None)
+            if meta is not None:
+                result = f(meta)
+                if result:
+                    return result
+
+    def __call__(self, typehint, **k):
+        return self._field_from_typehint(typehint, k)
+
+    def _get_type_hints(self, item):
+        hints = {}
+
+        for parent in item.__mro__[::-1]:
+            hints.update(get_type_hints(parent))
+
+        return hints
+
+    def convert(self, name, typehint, **kwargs):
+        opts = {**self._get_meta_options(name), **kwargs}
+        return self._field_from_typehint(typehint, opts)
+
+    def convert_all(self, type, ignore=frozenset()):
+        return {
+            k: self.convert(k, v)
+            for k, v in self._get_type_hints(type).items() if k not in ignore
+        }
+
+
+class AnnotationSchemaMeta(SchemaMeta):
+
+    @staticmethod
+    def __new__(mcls, name, bases, attrs, **k):
+        cls = super().__new__(mcls, name, bases, attrs)
+        meta = getattr(cls, 'Meta', None)
+        cls._register_as_scheme_for_target(meta)
+        return cls
+
+    @classmethod
+    def get_declared_fields(
+            mcls, klass, cls_fields, inherited_fields, dict_cls
+    ):
+        fields = super().get_declared_fields(
+            klass, cls_fields, inherited_fields, dict_cls
+        )
+
+        meta = getattr(klass, 'Meta', None)
+
+        if not meta:
+            return fields
+
+        target = getattr(meta, 'target', None)
+
+        if target is None:
+            return fields
+
+        converter_factory = getattr(meta, 'converter', BaseConverter)
```

Comment:
walk up MRO metas to find a converter if it's not present here. also don't do this in this method, but do it in `__new__`